common/src/main/java/tachyon/underfs/UnderFileSystem.java

```diff
@@ -364,6 +368,53 @@ public abstract OutputStream create(String path, short replication, int blockSiz
   public abstract String[] list(String path) throws IOException;
 
   /**
+   * Returns an array of strings naming the files and directories in the directory denoted by this
+   * abstract pathname, and all of its subdirectories.
+   *
+   * <p>
+   * If this abstract pathname does not denote a directory, then this method returns {@code null}.
+   * Otherwise an array of strings is returned, one for each file or directory in the directory and
+   * its subdirectories. Names denoting the directory itself and the directory's parent directory
+   * are not included in the result. Each string is a path relative to the given directory.
+   *
+   * <p>
+   * There is no guarantee that the name strings in the resulting array will appear in any specific
+   * order; they are not, in particular, guaranteed to appear in alphabetical order.
+   *
+   * @param path the path to list
+   * @return An array of strings naming the files and directories in the directory denoted by this
+   *         abstract pathname and its subdirectories. The array will be empty if the directory is
+   *         empty. Returns {@code null} if this abstract pathname does not denote a directory, or
+   *         if an I/O error occurs.
+   * @throws IOException if a non-Tachyon error occurs
+   */
+  public String[] listRecursive(String path) throws IOException {
+    // Clean the path by creating a URI and turning it back to a string
+    TachyonURI uri = new TachyonURI(path);
+    path = uri.toString();
+    List<String> returnPaths = new ArrayList<String>();
+    // Anything in this queue should have already been added to the returnPaths list if we're adding
+    // it (so the root doesn't get added)
```

Comment:
We need to call `list(path)` at the start to make sure `path` actually exists, and return null if it doesn't.